src/Cardano/Wallet/CoinSelection/LargestFirst.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE RankNTypes #-}
+
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- This module contains the implementation of largestFirst
+-- input selection algorithm
+
+
+module Cardano.Wallet.CoinSelection.LargestFirst (
+    largestFirst
+  ) where
+
+import Prelude
+
+import Cardano.Wallet.CoinSelection
+    ( CoinSelection (..), CoinSelectionError (..), CoinSelectionOptions (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), TxIn, TxOut (..), UTxO (..), balance )
+import Control.Monad
+    ( foldM, when )
+import Control.Monad.IO.Class
+    ( MonadIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), throwE )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Ord
+    ( comparing )
+
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadIO m
+    => CoinSelectionOptions
+    -> UTxO
+    -> NonEmpty TxOut
+    -> ExceptT CoinSelectionError m CoinSelection
+largestFirst opt utxo txOutputs = do
+    let txOutputsSorted = NE.toList $ NE.sortBy (comparing coin) txOutputs
+    let n = fromIntegral $ maximumNumberOfInputs opt
+    let nLargest = take n . L.sortBy (flip $ comparing (coin . snd)) . Map.toList . getUTxO
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
+    let utxoBalance = fromIntegral $ balance utxo
+    let numberOfUtxoEntries = fromIntegral $ L.length $ (Map.toList . getUTxO) utxo
+    let numberOfTransactionOutputs = fromIntegral $ NE.length txOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ NotEnoughMoney utxoBalance moneyRequested
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    -- FIXME ? we need to check if the transaction outputs are not redeemable
+
+    case foldM atLeast (L.reverse $ nLargest utxo, mempty) txOutputsSorted of
```

Comment:
Why is the list reversed here :thinking: ?

Reply:
we start to consume utxo from the smallest cut obtained by maximumNumberInputs (which takes those numbers of biggest utxos entries) . taken from : https://github.com/input-output-hk/cardano-wallet-legacy/blob/develop/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs#L215